File: Backtracking/Subsets.java
package Backtracking;

import java.util.ArrayList;
import java.util.List;

public class Subsets {
  List<List<Integer>> subsetsList;
  int[] nums;

  public List<List<Integer>> subsets(int[] nums) {
    subsetsList = new ArrayList<>();
    this.nums = nums;
    backtrack(new ArrayList<>(), 0);
    return subsetsList;
  }

  private void backtrack(List<Integer> curr, int i) {
    subsetsList.add(new ArrayList<>(curr));

    for (int j = i; j < nums.length; j++) {
      curr.add(nums[j]);
      backtrack(curr, j + 1);
      curr.remove(curr.size() - 1);
    }
  }
}

// Time O(n⋅2^n)
// Space O(n⋅2^n)

File: Backtracking/subsets.js
const subsets = (nums) => {
  const subsetsArr = [];

  const backtrack = (curr, i) => {
    subsetsArr.push([...curr]);
    for (i; i < nums.length; i++) {
      curr.push(nums[i]);
      backtrack(curr, i + 1);
      curr.pop();
    }
  };

  backtrack([], 0);
  return subsetsArr;
};

// Time O(n⋅2^n)
// Space O(n⋅2^n)

const nums = [1, 2, 3];
console.log(subsets(nums));